### PR TITLE
Resource list page with drafts and submitted for review

### PR DESCRIPTION
**Pull request by adithyadinesh0412**

### CODE REVIEW

File: src/database/migrations/20240612100550-add-new-status-to-resources-status-enum.js [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+
+/** @type {import('sequelize-cli').Migration} */
+module.exports = {
+	async up(queryInterface, Sequelize) {
+		const transaction = await queryInterface.sequelize.transaction()
+		try {
+			// Step 1: Create a new ENUM type with the new values
+			await queryInterface.sequelize.query(
+				`
+        CREATE TYPE "enum_resources_status_new" AS ENUM('DRAFT', 'SUBMITTED', 'IN_REVIEW', 'APPROVED', 'REJECTED', 'PUBLISHED', 'REJECTED_AND_REPORTED', 'REQUESTED_FOR_CHANGES');
```
Comment (priyanka-TL) [inline src/database/migrations/20240612100550-add-new-status-to-resources-status-enum.js]: @adithyadinesh0412 REQUESTED_FOR_CHANGES wont come in resource table

File: src/constants/common.js [hunk 2]
```diff
@@ -67,6 +67,30 @@ module.exports = {
...
+	PAGE_STATUS_VALUES: {
+		drafts: ['DRAFT'],
+		up_for_review: ['in_review'],
+		submitted_for_review: [
+			'IN_REVIEW',
+			'SUBMITTED',
+			'PUBLISHED',
+			'REJECTED',
+			'REJECTED_AND_REPORTED',
+			'REQUESTED_FOR_CHANGES',
```
Comment (adithyadinesh0412) [inline src/constants/common.js]: remove this

File: src/constants/common.js [hunk 3]
```diff
@@ -67,6 +67,30 @@ module.exports = {
 	SORT_ORDER: 'sort_order',
 	URL: 'url',
 	FILE_PATH: 'filePath',
+	PAGE_STATUS: 'page_status',
+	PAGE_STATUS_VALUES: {
+		drafts: ['DRAFT'],
+		up_for_review: ['in_review'],
```
Comment (adithyadinesh0412) [inline src/constants/common.js]: make it caps

File: src/database/migrations/20240612100550-add-new-status-to-resources-status-enum.js (hunk 1 above)
Comment (adithyadinesh0412) [inline src/database/migrations/20240612100550-add-new-status-to-resources-status-enum.js]: REQUESTED_FOR_CHANGES - remove

File: src/database/migrations/20240612102518-update-reviews-composite-key.js [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+					'INPROGRESS',
+					'REQUESTED_FOR_CHANGES',
+					'APPROVED',
+					'REJECTED',
+					'PUBLISHED'
+				),
+				defaultValue: 'NOT_STARTED',
+			},
+			organization_id: {
+				allowNull: false,
```
Comment (adithyadinesh0412) [inline src/database/migrations/20240612102518-update-reviews-composite-key.js:17]: add only the changes values

File: src/database/models/resources.js [hunk 5]
```diff
@@ -18,7 +18,16 @@ module.exports = (sequelize, DataTypes) => {
...
-				type: DataTypes.ENUM('DRAFT', 'SUBMITTED', 'IN_REVIEW', 'APPROVED', 'REJECTED', 'PUBLISHED'),
+				type: DataTypes.ENUM(
+					'DRAFT',
+					'SUBMITTED',
+					'IN_REVIEW',
+					'APPROVED',
+					'REJECTED',
+					'PUBLISHED',
+					'REJECTED_AND_REPORTED',
+					'REQUESTED_FOR_CHANGES'
```
Comment (adithyadinesh0412) [inline src/database/models/resources.js]: REQUESTED_FOR_CHANGES - remove

File: src/database/models/reviews.js [hunk 6]
```diff
@@ -9,12 +9,13 @@ module.exports = (sequelize, DataTypes) => {
 			},
 			resource_id: {
 				allowNull: false,
-				primaryKey: true,
 				type: DataTypes.INTEGER,
+				primaryKey: true, // Part of composite primary key
```
Comment (adithyadinesh0412) [inline src/database/models/reviews.js:13]: recheck

File: src/database/queries/resourcesCreatorMapping.js [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+	} catch (error) {
+		return error
+	}
+}
+
+exports.findAll = async (filter, attributes = {}) => {
+	try {
+		const res = await resource_user_mapping.findAll({
+			where: filter,
+			attributes,
```
Comment (adithyadinesh0412) [inline src/database/queries/resourcesCreatorMapping.js:16]: check for options

File: src/database/queries/reviews.js [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+			attributes,
+			raw: true,
+		})
+
+		return res
+	} catch (error) {
+		return error
+	}
+}
+exports.countDistinct = async (filter, attributes = {}) => {
```
Comment (adithyadinesh0412) [inline src/database/queries/reviews.js:17]: Return the unique count of resource id. Also check if we can group and count distinct

File: src/locales/en.json [hunk 9]
```diff
@@ -61,6 +61,7 @@
 	"CERTIFICATE_FETCH_FAILED": "Error in fetching Certificate",
 	"PROJECT_CREATED_SUCCESSFULLY": "Project created successfully",
 	"PROJECT_UPDATED_SUCCESSFUL": "Project updated successfully",
+	"FORBIDDEN_RESOURCE_UPDATE": "Forbidden Update Operation",
```
Comment (adithyadinesh0412) [inline src/locales/en.json]: Change the error message.

File: src/services/projects.js [hunk 10]
```diff
@@ -27,6 +30,12 @@ module.exports = class ProjectsHelper {
 				updated_by: loggedInUserId,
 			}
 			let projectCreate = await resourceQueries.create(projectData)
```
Comment (adithyadinesh0412) [inline src/services/projects.js:45]: Add , review type in resource table while creating. Fetch from org_ext or env

File: src/services/projects.js [hunk 11]
```diff
@@ -47,6 +56,48 @@ module.exports = class ProjectsHelper {
...
 	static async update(resourceId, orgId, loggedInUserId, bodyData) {
 		try {
+			let blockUpdate = false
+			const forbidden_resource_statuses = [
+				common.RESOURCE_STATUS_PUBLISHED,
+				common.RESOURCE_STATUS_REJECTED,
+				common.RESOURCE_STATUS_REJECTED_AND_REPORTED,
+				common.RESOURCE_STATUS_SUBMITTED,
+				common.RESOURCE_STATUS_APPROVED,
+			]
```
Comment (adithyadinesh0412) [inline src/services/projects.js:79]: get the org id from the mapping table

Comment (adithyadinesh0412) [reply src/services/projects.js:79]: pass status in fetch it self.

File: src/services/projects.js [hunk 12]
```diff
@@ -47,6 +56,48 @@ module.exports = class ProjectsHelper {
...
+			const fetchResource = await resourceQueries.findOne({
+				id: resourceId,
+				organization_id: orgId,
+			})
+			const countReviews = await reviewsQueries.countDistinct({
```
Comment (adithyadinesh0412) [inline src/services/projects.js:87]: check count only if its in review

File: src/services/projects.js [hunk 13]
```diff
@@ -85,9 +136,13 @@ module.exports = class ProjectsHelper {
 				let filter = {
 					id: resourceId,
 					organization_id: orgId,
+					status: {
+						[Op.notIn]: forbidden_resource_statuses,
```
Comment (adithyadinesh0412) [inline src/services/projects.js]: No need to check here - status

File: src/services/resource.js [hunk 14]
```diff
@@ -34,20 +36,80 @@ module.exports = class resourceHelper {
...
+				const changesCount = await reviewsQueries.countDistinct({
+					organization_id: {
+						[Op.in]: OrganizationIds,
+					},
+					resource_id: {
+						[Op.in]: uniqueResourceIds,
+					},
+					status: common.REVIEW_STATUS_REQUESTED_FOR_CHANGES,
+				})
+				result.changes_requested_count = changesCount
```
Comment (adithyadinesh0412) [inline src/services/resource.js:89]: Add in API - doc

File: src/services/projects.js [hunk 15]
```diff
@@ -47,6 +56,48 @@ module.exports = class ProjectsHelper {
...
+				: fetchResource.status === common.RESOURCE_STATUS_IN_REVIEW && countReviews <= 0
+
+			if (blockUpdate) {
+				return responses.failureResponse({
+					message: 'FORBIDDEN_RESOURCE_UPDATE',
+					statusCode: httpStatusCode.bad_request,
+					responseCode: 'CLIENT_ERROR',
+				})
+			}
+
```
Comment (adithyadinesh0412) [inline src/services/projects.js:111]: Add categories how they send.

File: src/services/resource.js [hunk 16]
```diff
@@ -34,20 +36,80 @@ module.exports = class resourceHelper {
...
+					}
+
+					if (queryParams.status.split(',').includes(common.REVIEW_STATUS_REQUESTED_FOR_CHANGES)) {
+						// fetch the resource ids from review table - changes requested
+						let changeRequestedResources = await reviewsQueries.findAll(
+							{
+								resource_id: {
+									[Op.in]: uniqueResourceIds,
+								},
+								status: common.REVIEW_STATUS_REQUESTED_FOR_CHANGES,
```
Comment (adithyadinesh0412) [inline src/services/resource.js:107]: Pass orgid also. because its the DC

File: src/services/resource.js [hunk 17]
```diff
@@ -34,20 +36,80 @@ module.exports = class resourceHelper {
...
+							},
+							['resource_id']
+						)
+
+						changeRequestedResources = [...new Set([..._.map(changeRequestedResources, 'resource_id')])]
+						filter.id = {
+							[Op.in]: changeRequestedResources,
+						}
+						filter.status = common.RESOURCE_STATUS_IN_REVIEW
+					}
```
Comment (adithyadinesh0412) [inline src/services/resource.js:120]: add pagination

File: src/database/models/resources.js [hunk 18]
```diff
@@ -18,7 +18,15 @@ module.exports = (sequelize, DataTypes) => {
 			},
 			status: {
 				allowNull: false,
-				type: DataTypes.ENUM('DRAFT', 'SUBMITTED', 'IN_REVIEW', 'APPROVED', 'REJECTED', 'PUBLISHED'),
+				type: DataTypes.ENUM(
+					'DRAFT',
+					'SUBMITTED',
+					'IN_REVIEW',
+					'APPROVED',
```
Comment (priyanka-TL) [inline src/database/models/resources.js:25]: @adithyadinesh0412 we don't have approved status

**Pull request merged**